app/queue.py

```diff
@@ -105,20 +112,24 @@ def poll(self, count=10) -> tuple[UUID, list[str]]:
         receipt = uuid4()
         in_flight_key = Buffer.IN_FLIGHT.inflight_name(receipt, self._suffix)
         results = self.__move_to_inflight(in_flight_key, count)
+        put_batch_saving_in_flight_metric(1)
         return (receipt, results)
 
     def expire_inflights(self):
         args = [f"{Buffer.IN_FLIGHT.inflight_prefix()}:{self._suffix}*", self._inbox, self._expire_inflight_after_seconds]
         expired = self.scripts[self.LUA_EXPIRE_INFLIGHTS](args=args)
         if expired:
+            put_batch_saving_expiry_metric(len(expired))
             current_app.logger.warning(f"Moved inflights {expired} back to inbox {self._inbox}")
 
     def acknowledge(self, receipt: UUID):
         inflight_name = Buffer.IN_FLIGHT.inflight_name(receipt, self._suffix)
         self._redis_client.delete(inflight_name)
+        put_batch_saving_inflight_processed(1)
 
     def publish(self, message: str):
         self._redis_client.rpush(self._inbox, message)
+        put_batch_saving_metric(self, 1)
```

Comment:
Do we have a quota on how many calls we are allowed to send to AWS metrics? Thinking of the case where traffic will be high and many publish might happen per second.

Reply:
No concerns related to volume. CloudWatch embedded metrics is a two step process. Our app will log to a CloudWatch log group using a specific format. CloudWatch will detect this format and automatically generate the metrics async.

Reply:
It seems we are allowed 1,000,000 API calls per month. We'll probably make a couple thousands per month (if HC sends 40,000 every business days, that would be 800,000 per month for that specific service and only for the API call going through the publish buffer queue).

I was wondering more about if there is a rate limit on AWS side, how many times per second or minute we can hit them, considering the bursty traffic, it needs to be at least 6,000 a minute.

Reply:
From what i know that only applies to the standard CloudWatch metrics where you invoke `PutMetrics`.

Embedded metrics are intended for high volumes and i couldn't find any service limits related to the amount of data stored with the exception of a message size of `256Kb`. Majority of the cost with this service is the number of unique metrics you log.

At the moment we're generating `3` unique metrics for Batch savings.